pkg/kubelet/pleg/evented.go

```diff
@@ -0,0 +1,384 @@
+)
+
+// isEventedPLEGInUse indicates whether evented PLEG is in use. Even after enabling
+// the Evented PLEG feature gate, there could be several reasons it may not be in use.
+// e.g. Streaming data issues from the runtime or the runtime does not implement the
+// container events stream.
+var isEventedPLEGInUse = false
+
+type EventedPLEG struct {
+	// The container runtime.
+	runtime kubecontainer.Runtime
+
+	// The runtime service
+	runtimeService internalapi.RuntimeService
+
+	// The channel from which the subscriber listens events.
+	eventChannel chan *PodLifecycleEvent
+
+	// Cache for storing the runtime states required for syncing pods.
+	cache kubecontainer.Cache
+
+	// For testability.
+	clock clock.Clock
+
+	// GenericPLEG is used to force relist when required
+	genericPleg PodLifecycleEventGenerator
+
+	// The maximum number of retries when getting container events from the runtime.
+	eventedPlegMaxStreamRetries int
+
+	// The period for relisting all containers using generic pleg as a fallback
+	genericPlegRelistPeriod time.Duration
+
+	// The threshold value for the Generic PLEG health
+	genericPlegRelistThreshold time.Duration
+
+	// Stop the Evented PLEG by closing the channel.
+	stopCh chan struct{}
+
+	// Stops the periodic update of the cache global timestamp
+	stopCacheUpdateCh chan struct{}
+
+	// Locks the start/stop operation of the Evented PLEG
+	runningMu sync.Mutex
+}
+
+// NewEventedPLEG instantiates a new EventedPLEG object and return it.
+func NewEventedPLEG(runtime kubecontainer.Runtime, runtimeService internalapi.RuntimeService, eventChannel chan *PodLifecycleEvent,
+	cache kubecontainer.Cache, genericPleg PodLifecycleEventGenerator, eventedPlegMaxStreamRetries int,
+	genericPlegRelistPeriod time.Duration, genericPlegRelistThreshold time.Duration, clock clock.Clock) PodLifecycleEventGenerator {
+	return &EventedPLEG{
+		runtime:                     runtime,
+		runtimeService:              runtimeService,
+		eventChannel:                eventChannel,
+		cache:                       cache,
+		genericPleg:                 genericPleg,
+		eventedPlegMaxStreamRetries: eventedPlegMaxStreamRetries,
+		genericPlegRelistPeriod:     genericPlegRelistPeriod,
+		genericPlegRelistThreshold:  genericPlegRelistThreshold,
+		clock:                       clock,
+	}
+}
+
+// Watch returns a channel from which the subscriber can receive PodLifecycleEvent events.
+func (e *EventedPLEG) Watch() chan *PodLifecycleEvent {
+	return e.eventChannel
+}
+
+// Relist relists all containers using GenericPLEG
+func (e *EventedPLEG) Relist() {
+	e.genericPleg.Relist()
+}
+
+// Start starts the Evented PLEG
+func (e *EventedPLEG) Start() {
+	e.runningMu.Lock()
+	defer e.runningMu.Unlock()
+	if !isEventedPLEGInUse {
+		isEventedPLEGInUse = true
+		e.stopCh = make(chan struct{})
+		e.stopCacheUpdateCh = make(chan struct{})
+		go wait.Until(e.watchEventsChannel, 0, e.stopCh)
+		go wait.Until(e.updateGlobalCache, 5*time.Second, e.stopCacheUpdateCh)
```

Comment:
define a constant for `5*time.Second`

Reply:
Defined the constant: `globalCacheUpdatePeriod`.